File: autoping.php
<?
	define('NO_WEB',1);
	define('NUM',20);
	define('PATH_TO_ROOT','./');
	include PATH_TO_ROOT."conf.php";

function ping_val($adr){
	if(PLATFORM=='windows'){
		$r = `ping -n 1 -w 1000 $adr`;
		$m = preg_match('/time=(\d+)ms/',$r,$ma);
		if($m)
			return $ma[1];
		return -1;
	}else{
		$r = `ping -s 32 -c 1 -w 1 $adr 2> /dev/null`;
		$m = preg_match('/time=(\d+)(?:\.(\d+))? ms/',$r,$ma); // тут косяк. Парень считает, что время всегда с плавающей точкой. Иногда же мы получаем время целым числом
		if($m){
			if(strpos($ma[1],'.')===false)
				$ma[1] .= '.0';
			return $ma[1];
		}
		return -1;
	}
}
function try_register(){
	global $db,$THIS_NUM;
	@$db->Query('insert into monitor_5min_ins values (0,'.$THIS_NUM.',0)');
	if($db->AffectedRows()==0){
		echo "i'm a second copy of ".$THIS_NUM.".\n";
		exit;
	}
}

	if(!$db->Connect())
		exit;

	if(!isset($argv[1])){ // если не fork
		echo "BEGIN;#######################".date("Y-m-d H:i:s")."#######################\n";
		$THIS_NUM=0;
		try_register();

		$r = $db->GetRow('
			select
				max(time300) as tmax,
				FLOOR(UNIX_TIMESTAMP()/300) as tcur
			from
				monitor_5min
			where
				ip_int!=0
		');	//min(time300) as tmin,

		if($r['tmax']==$r['tcur']) // если не прошло 5 минут с последнего запуска - завершаем работу скрипта
			die("Wait some minutes plz. max(time300)=".$r['tmax']);

		$T=$r['tcur'];

		if(($T % 288) == 240){ // если остаток от деления текущего timestamp на 288 равен 4м минутам - выполняется кусок этого кода.
			echo "Summarizing...\n";
			$t = $T-288; //($T%288)-288;
			$A = $db->GetRow('
				select
					max(time3600) as A
				from
					monitor_1h
			');
			$db->Query('
				insert into
					monitor_1h
				select
					ip_int,
					FLOOR(time300/12) as time3600,
					sum(IF(value=0,1,0)) + 12 - count(*) as bad_count,
					sum(value) as good_sum
				FROM
					monitor_5min
				where
					ip_int>0
				and
					time300<'.$t.'
				GROUP BY
					ip_int,
					time3600
			');
			$db->Query('
				delete from
					monitor_5min
				where
					ip_int > 0
				and
					time300 < '.$t
			);
			$db->Query('truncate table monitor_ips');
			$db->Query('update monitor_clients set period_use=0');
		}

		$r = $db->GetRow('
			select
				value
			from
				monitor_5min
			where
				ip_int = 0
			and
				time300 = 0
			limit 1
		');

		if(!$r){
			$db->Query('insert into monitor_5min (ip_int,time300,value) values(0,0,1)');
			$ITERATION=1;
		}else{
			$db->Query('update monitor_5min set value=value+1 where ip_int=0 and time300=0');
			$ITERATION=$r['value']+1;
		}
		echo "Iteration: ".$ITERATION."\n";

		$db->Query('truncate monitor_5min_ins');
		/**
		 * запускаем 20 форков себя, которые будут что-то делать параллельно,
		 * передавая им 2 аргумента:
		 * 1 - порядковый номер форка (в форке это - $THIS_NUM)
		 * 2 - время текущего вычисления (в форке это - $T)
		 *
		 * вывод форков перенаправляется в /dev/null
		 */
		for($i=1;$i<NUM;$i++){
			if(PLATFORM=='windows'){
				system("php -f ".PATH_TO_ROOT."autoping.php $i $T");
			}else{
				if(SERVER=="tiberis"){
					system("php -c /etc/ -f ".PATH_TO_ROOT."autoping.php $i $T > /dev/null &");
				}else{
					system("php -f ".PATH_TO_ROOT."autoping.php $i $T > /dev/null &");
				}
			}
		}
	}else{ // если текущий процесс - fork
		$ITERATION=-1;
		$THIS_NUM = $argv[1];
		try_register();
		$T=$argv[2];
		if(!$T)
			return;
	}

	$R=array();
	$db->Query( // выборка клиентов и их подсетей
		'select
			usage_ip_routes.net,
			usage_ip_ports.client
		from
			usage_ip_routes
		INNER JOIN
			usage_ip_ports
		ON
			usage_ip_ports.id = usage_ip_routes.port_id
		where
			usage_ip_routes.actual_from <= NOW()
		and
			usage_ip_routes.actual_to > NOW()
	');
	while($r=$db->NextRecord()) // выборка первых 2х(или, если подсеть только из одного ip - тогда 1) ip'шников сети в массив $R
		add_ip($R,$r[0],'');

	$db->Query( // выборка клиентов и их ip адресов с ppp подключениями
		'select
			ip,
			client
		from
			usage_ip_ppp
		where
			actual_from <= NOW()
		and
			actual_to >= NOW()
	');
	while($r=$db->NextRecord()) // помещение ip адресов в массив $R
		add_ip($R,$r[0],'');

	$db->Query( // выботка ip адресов и клиентов
		'select
			ip,
			client
		from
			tech_cpe
		where
			actual_from <= NOW()
		and
			actual_to >= NOW()
		and
			ip != ""
	');
	while($r=$db->NextRecord()) // помещение в массив $R ip адреса и клиента, им владеющего
		add_ip($R,$r[0],$r[1]);

	$db->Query( // выборка подсетей и роутеров, которым они принадлежат
		'select
			net,
			router
		from
			tech_routers
		where
			actual_from <= NOW()
		and
			actual_to >= NOW()
	');
	while($r=$db->NextRecord()) // добавление в массив $R ip адресов и роутеров
		add_ip($R,$r[0],'mcn.'.$r[1]);

	$IPs_bad = array();
	$db->Query('select INET_NTOA(ip_int) as ip from monitor_ips'); // выборка ip адресов, по которым был обнаружен процент потерь
	while($r=$db->NextRecord())
		$IPs_bad[$r['ip']]=$r['ip'];

	$clients=array();
	$db->Query('select id,client from monitor_clients'); // выборка клиентов, за которыми ведется мониторинг на email
	while($r=$db->NextRecord())
		$clients[$r['client']]=$r['id'];

	$cnt_thread=0;
	$cnt_total=0;

	ksort($R); // сортировка массива по ключам(ip адресам), сохраняя связи между ключами и значениями
	foreach($R as $ip=>$client){
		if(($cnt_total % NUM) == $THIS_NUM){// если остаток от деления на количество форков равняется идентификатору текущего форка - обрабатываем, иначе пропускаем
			$v=ping_val($ip); // пингуем ip адрес системно (system("ping...")). Если пропинговался - $v = количеству милисекунд пинга, если нет = -1

			if($v<0)
				$v=0;
			elseif($v==0)
				$v=1;

			if(isset($IPs_bad[$ip])){ // если по этому ip адресу был обнаружен процент потерь
				if($v){
//					$db->Query('delete from monitor_ips where ip_int=INET_ATON("'.$ip.'")');
				}else{
					$db->Query( // увеличиваем количество потерь на 1
						'update
							monitor_ips
						set
							count=count+1
						where
							ip_int = INET_ATON("'.$ip.'")
					');
				}
			}elseif(!$v && isset($clients[$client])){ // если по ip адресу еще не было обнаружено потерь, но это 1я, и за клиентом ведется мониторинг - добавляем в таблицу с плохими ip адресами
				$db->Query('
					insert into
					monitor_ips (
						ip_int,
						monitor_id,
						count
					)
					values (
						INET_ATON("'.$ip.'"),
						'.$clients[$client].',
						1
					)
				');
			}
			$db->Query( // наполняем таблицу статистики. поле value - это время пинга
				'insert into
				monitor_5min_ins (
					ip_int,
					time300,
					value
				)
				values (
					INET_ATON("'.$ip.'"),
					'.$T.',
					'.$v.'
				)
			');
			$cnt_thread++;
		}
		$cnt_total++;
	}
	@$db->Lock('monitor_5min_ins');
	$db->Query( // удаляем текущий форк из таблицы - этим разрешаем форкать скрипт, с текущим порядковым номером, в будущем
		'delete from
			monitor_5min_ins
		where
			ip_int=0
		and
			time300='.$THIS_NUM
	);
	$C=$db->GetRow( // выбираем из таблицы форков, которые запущены в настоящее время.
		'select
			count(*) as C
		from
			monitor_5min_ins
		where
			ip_int=0
		and
			value=0
	');
	@$db->Unlock();
	if($C['C']==0){ // блок finally - если текущий процесс последний из всех форков - выполняем
		$db->Query( // перемещаем все записи статистики в боевую таблицу
			'insert into
				monitor_5min
			select
				*
			from
				monitor_5min_ins
		');
		$db->Query('truncate monitor_5min_ins'); // подчищаем за собой таблицу с временной статистикой

		$M=$db->AllRecords( // получаем клиентов, на которых выставлены флаги мониторинга.
			'select
				monitor_clients.*
			from
				monitor_clients
			inner join
				monitor_ips
			ON
				monitor_ips.monitor_id = monitor_clients.id
			where
				monitor_ips.count > monitor_clients.allow_bad
			group by
				monitor_clients.id
		');
		foreach($M as $r){
			if($r['period_use']<=0){ // если пришло время слать спам, тогда отправляем статистику по этому клиенту
				$r['period_use'] = $r['period_mail'];
				$body = "мЕСДЮВМШЕ ОХМЦХ С ЙКХЕМРЮ ".$r['client']." (ПЮГПЕЬЕМН - ".$r['allow_bad']."):\n"; // "Неудачные пинги у клиента ".$r['client']." (разрешено - ".$r['allow_bad']."):\n"
				$body = "IP \tЙНКХВЕЯРБН МЕСДЮВМШУ ОХМЦНБ\n"; // "IP \tколичество неудачных пингов\n"
				$db->Query( // выбираем ip адрес и количество потерь, на конкретного клиента
					'select
						INET_NTOA(ip_int) as ip,
						count
					from
						monitor_ips
					where
						monitor_id='.$r['id']
				);
				while($r2=$db->NextRecord()){
					$body .= $r2['ip']."\t".$r2['count']."\n"; // ip \t количество
				}

				$r['email'] = str_replace(';',',',$r['email']);
				if(
					(
						defined('MAIL_TEST_ONLY')
					&&
						(MAIL_TEST_ONLY==1)
					)
				||
					$is_test
				) $r['email']='dnsl48@gmail.com';//'shepik@yandex.ru, mak@mcn.ru';

				$headers  = "From: MCN Info <info@mcn.ru>\n";
				$headers .= "Content-Type: text/plain; charset=windows-1251\n";
				$subj = "лНМХРНПХМЦ: ЙКХЕМР ".$r['client']; // Мониторинг: клиент ".$r['client'];
				if($r['email'])
					mail($r['email'],$subj,$body,$headers);
			}else
				$r['period_use']--;
			
			$db->Query('
				update
					monitor_clients
				set
					period_use = '.$r['period_use'].'
				where
					id='.$r['id']
			);
		}
	}
	echo "Processed: ".$cnt_total." (this thread - ".$cnt_thread.")\n";
	echo "END;#########################".date("Y-m-d H:i:s")."#######################\n\n\n";
?>

File: stat/autoping.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__THREADS = 60	#количество потоков
__MAX_SQL_LEN = 800000 #максимальная длина sql запроса
__LOCKFILE = 'autoping.lock' #файл блокировок

__MAIL_FROM = 'MCN Info <info@mcn.ru>'
__MAIL_FROM_RAW = 'info@mcn.ru'
__MAIL_SMTP_SERV = 'localhost'
__MAIL_SMTP_PORT = 25

__DEBUG_MODE = False #если True, вместо insert'ов в базу будет вывод запросов
__DEBUG_SWAP_TABLES_MODE = False #если True, insert'ы пишутся в другую таблицу. Дополняет предыдущую директиву __DEBUG_MODE
__DEBUG_FORCE_MAIL = False #если True, письмо отправляется на email, даже если период отправления не настал
__DEBUG_MAIL = False #если True, письмо со спамом отправится на __DEBUG_ADDR вместо оригинального мейла, указанного в базе
__DEBUG_ADDR = 'dga@mcn.ru' #смотреть выше

import time
import threading
import Queue
import MySQLdb
import smtplib
import subprocess
import re
import fcntl
import os.path
import base64

v = {'host':'localhost','user':'stat_operator','passwd':'3616758a','db':'nispd','port':3306}
t = {'host':'127.0.0.1','user':'stat_operator','passwd':'3616758a','db':'nispd','port':3307}
c = v

connection = MySQLdb.connect(
	host=c['host'],
	user=c['user'],
	passwd=c['passwd'],
	db=c['db'],
	port=c['port']
)

ping_regex = re.compile("time=(\d+)(?:\.(\d+))? ms")
cidr_regex = re.compile("(\d+)\.(\d+)\.(\d+)\.(\d+)(?:\/(\d+))?")

def send_spam():
	"""Отправляет отчет по потерям на email, для клиент-email связки, указанной
	в monitor_clients"""

	smtp = smtplib.SMTP(__MAIL_SMTP_SERV,__MAIL_SMTP_PORT)

	cur = connection.cursor()
	cur.execute('''
		select
			mc.id,
			mc.client,
			mc.email,
			mc.allow_bad,
			mc.period_mail,
			mc.period_use
		from
			monitor_clients mc
		inner join
			monitor_ips mi
		ON
			mi.monitor_id = mc.id
		where
			mi.count > mc.allow_bad
		group by
			mc.id
	''')
	while True:
		row = cur.fetchone()
		if row == None:
			break
		nperiod = row[5]
		if __DEBUG_FORCE_MAIL or row[5] <= 0:
			msg = "From: %s\r\nTo: %s\r\n"
			msg += "Subject: =?utf-8?B?%s?=\r\n"%base64.b64encode('неудачный ping'.decode('koi8_r').encode('utf8'))
			msg += 'Content-Type: text/plain; charset = "koi8-r"\r\n'
			msg += "Content-Transfer-Encoding: 8bit\r\n"
			msg += "\r\n"
			msg += "Неудачные пинги у клиента %s (разрешено - %s):\n" % (row[1],row[3])# $r['client'], $r['allow_bad']
			msg += "IP \tколичество неудачных пингов\n"

			cur1 = connection.cursor()
			cur1.execute('''
				select
					inet_ntoa(ip_int),
					`count`
				from
					monitor_ips
				where
					monitor_id=%i
			''' % row[0])
			while True:
				row1 = cur1.fetchone()
				if row1 == None:
					break
				msg += "%s\t%s\n" % (str(row1[0]),str(row1[1]))
			cur1.close()
			maddrs = row[2].split(';') if not __DEBUG_MAIL else __DEBUG_ADDR.split(',')

			smtp.sendmail(__MAIL_FROM_RAW,maddrs,msg%(__MAIL_FROM,','.join(maddrs)))

			print "Mail sent on ", ','.join(maddrs)
		else:
			nperiod -= 1

		q = '''
			update
				monitor_clients
			set
				period_use = %i
			where
				id = %i
		''' % (nperiod,row[0])
		if not __DEBUG_MODE:
			cur.execute(q)
		else:
			print q
	cur.close()
	smtp.quit()

def get_flock(curtime):
	"""Читает файл блокировки. Если другой скрипт уже пингует полученную пятиминутку - возвращает False,
	иначе записывает в файл идентификатор своей пятиминутки, сохраняя остальные идентификаторы файла."""
	opmode = 'a+' if os.path.exists(__LOCKFILE) else 'w+'
	fl = False
	f = open(__LOCKFILE,opmode)
	fcntl.lockf(f,fcntl.LOCK_EX)
	f.seek(0)
	works = f.read().split(',')
	if not str(curtime) in [str(i) for i in works if works]:
		f.seek(0)
		f.truncate(0)
		works.append(curtime)
		f.write(','.join([str(i) for i in works if i]))
		f.flush()
		fl = True
	fcntl.lockf(f,fcntl.LOCK_UN)
	f.close()
	return fl

def release_flock(curtime):
	"""Открывает файл блокировки, вычищает из него свой идентификатор пятиминутки, оставляя
	чужие идентификаторы нетронутыми."""
	f = open(__LOCKFILE, 'a+')
	fcntl.lockf(f,fcntl.LOCK_EX)
	f.seek(0)
	works = f.read().split(',')
	works_buf = []
	for i in [str(n) for n in works if n]:
		if int(i) != curtime:
			works_buf.append(i)
	f.seek(0)
	f.truncate(0)
	f.write(','.join([str(i) for i in works_buf if i]))
	f.flush()
	fcntl.lockf(f,fcntl.LOCK_UN)
	f.close()

def set_result():
	"""Читает глобальный(для модуля) список Stats, в котором должны находиться данные
	пропинговки. Составляет из данных, и возвращает sql запрос для вставки данных в базу."""
	sql_begin = 'insert into monitor_5min (ip_int,time300,value) values '
	if __DEBUG_SWAP_TABLES_MODE:
		sql_begin = 'insert into monitor_5min_test (ip_int,time300,value) values '
	ins = "(inet_aton('%s'),%s,%s),"
	sql = ''
	ret = []

	for p in Stats:
		if not len(sql):
			sql = sql_begin + ins % (str(p[0]),str(p[1]),str(p[2]))
			if len(sql)-1 > __MAX_SQL_LEN:
				raise "Don't have enough space for sql query..."
			else:
				continue
		if len(sql + ins % (str(p[0]),str(p[1]),str(p[2])))-1 > __MAX_SQL_LEN:
			ret.append(sql[0:-1])
			sql = sql_begin + ins % (str(p[0]),str(p[1]),str(p[2]))
			continue
		sql += ins % (str(p[0]),str(p[1]),str(p[2]))
	ret.append(sql[0:-1])

	for i in IPs_bad.keys():
		if IPs_bad[i][2]:
			if __DEBUG_SWAP_TABLES_MODE:
				print '''
					insert into
						monitor_ips
					set
						ip_int = inet_aton('%s'),
						monitor_id = %i,
						`count` = %i
					on duplicate key update
						`count` = values(`count`)
				''' % (i,IPs_bad[i][0],IPs_bad[i][1])
			else:
				ret.append('''
					insert into
						monitor_ips
					set
						ip_int = inet_aton('%s'),
						monitor_id = %i,
						`count` = %i
					on duplicate key update
						`count` = values(`count`)
				''' % (i,IPs_bad[i][0],IPs_bad[i][1]))

	return ret


def get_source():
	"""Читает из базы ip адресы, которые необходимо пропинговать, и возвращает список
	кортежей вида (список ip адресов, идентификатор клиента). Так же заполняет
	глобальные(для модуля) списки IPs_bad("плохие" ip, на которые были обнаружены потери сегодня)
	и mon_cli(список клиентов, за которыми ведется мониторинг потерь, с отправкой на email)"""
	nets = []
	cur = connection.cursor()

	cur.execute('''
		select
			trim(ip),
			client
		from
			usage_welltime
		where
			actual_from <= NOW()
		and
			actual_to > NOW()
		and
			ip != ""
	''')
	row = True
	while row:
		row = cur.fetchone()
		if row == None:
			break
		if not row[0]:
			continue
		nets.append((from_cidr(row[0]),row[1]))


	cur.execute('''
		select
			trim(usage_ip_routes.net),
			usage_ip_ports.client
		from
			usage_ip_routes
		INNER JOIN
			usage_ip_ports
		ON
			usage_ip_ports.id = usage_ip_routes.port_id
		where
			usage_ip_routes.actual_from <= NOW()
		and
			usage_ip_routes.actual_to > NOW()
	''')
	row = True
	while row:
		row = cur.fetchone()
		if row == None:
			break
		if not row[0]:
			continue
		nets.append((from_cidr(row[0]),None))

	cur.execute('''
		select
			trim(ip),
			client
		from
			usage_ip_ppp
		where
			actual_from <= NOW()
		and
			actual_to >= NOW()
	''')
	row = True
	while row:
		row = cur.fetchone()
		if row == None:
			break
		if not row[0]:
			continue
		nets.append((from_cidr(row[0]),None))

	cur.execute('''
		select
			trim(ip),
			client
		from
			tech_cpe
		where
			actual_from <= NOW()
		and
			actual_to >= NOW()
		and
			ip != ""
	''')
	row = True
	while row:
		row = cur.fetchone()
		if row == None:
			break
		if not row[0]:
			continue
		nets.append((from_cidr(row[0]),row[1]))

	cur.execute('''
		select
			trim(net),
			router
		from
			tech_routers
		where
			actual_from <= NOW()
		and
			actual_to >= NOW()
	''')
	row = True
	while row:
		row = cur.fetchone()
		if row == None:
			break
		if not row[0]:
			continue
		nets.append((from_cidr(row[0]),'mcn'+str(row[1])))

	cur.execute('select inet_ntoa(ip_int), monitor_id, `count` from monitor_ips')
	global IPs_bad
	while True:
		row = cur.fetchone()
		if row == None:
			break
		IPs_bad[row[0]] = [row[1],row[2],False]

	cur.execute('select id,client from monitor_clients')
	global mon_cli
	while True:
		row = cur.fetchone()
		if row == None:
			break
		mon_cli[row[1]] = row[0]

	cur.close()

	ret = {}
	for i in nets:
		for j in i[0]:
			ret[j] = i[1]

	return ret

def from_cidr(addr):
	"""Получает cidr адрес, либо просто ip. Возвращает список либо первых 2 из подсети,
	либо с одним ip адресом, который был получен"""
	m = cidr_regex.match(addr)
	if not m:
		return []
	m = m.groups()
	if not m[4] or int(m[4])==32:
		return ["%i.%i.%i.%i" % (int(m[0]),int(m[1]),int(m[2]),int(m[3]))]
	else:
		return [
			"%i.%i.%i.%i" % (int(m[0]),int(m[1]),int(m[2]),int(m[3])+1),
			"%i.%i.%i.%i" % (int(m[0]),int(m[1]),int(m[2]),int(m[3])+2)
		]

def agg_1h(curtime):
	"""Аггрегирует статистику по пятиминуткам(5min), в статистику по часам(1h), подчищая таблицы
	monitor_5min и monitor_ips. Выставляет всем клиентам флаг period_use в 0(таблица monitor_clients)"""
	cur = connection.cursor()
	cur.execute('start transaction')
	cur.execute('''
		insert into
			monitor_1h
		select
			ip_int,
			FLOOR(time300/12) as time3600,
			sum(IF(value=0,1,0)) + 12 - count(*) as bad_count,
			sum(value) as good_sum
		FROM
			monitor_5min
		where
			ip_int>0
		and
			time300<%s
		GROUP BY
			ip_int,
			time3600
	''' % (curtime-288))

	cur.execute('''
		delete from
			monitor_5min
		where
			ip_int > 0
		and
			time300 < %s
	''' % (curtime-288))

	cur.execute('truncate table monitor_ips')
	cur.execute('update monitor_clients set period_use=0')

	cur.execute('commit')
	cur.close()

def check_curtime():
	"""Проверяет, не собрана ли уже статистика на текущую пятиминутку. Если уже собрана,
	возвращает 0, иначе возвращает идентификатор текущей пятиминутки."""
	cur = connection.cursor()
	sql = """
		select
			max(time300) as tmax,
			floor(unix_timestamp()/300) as tcur
		from
			monitor_5min
		where
			ip_int!=0
	"""
	cur.execute(sql)
	row = cur.fetchone()
	cur.close()
	if row[0] < row[1]:
		return row[1]
	else:
		return 0

def ping_addr(addr):
	"""Отправляет пинг на полученный ip адрес, возвращает распарсенный результат в виде числа.
	Если прошла потеря, возвращает -1"""
	msg = subprocess.Popen("ping -s 32 -c 1 -w 1 %s 2> /dev/null" % str(addr),shell=True,stdout=subprocess.PIPE).stdout.read()
	m = ping_regex.findall(msg)
	if len(m) and len(m[0]):
		return int(m[0][0])
	return -1

def pinger_daemon(idx, nowtime, strtime):
	"""Демон, ожидающий очереди ip адреса. Получает его, пингует, пишет по нему статистику, и ждет
	очереди дальше."""
	global IPs_bad
	global threads_processing_cnt
	print "Starting Thread #%i"%idx
	while True:
		item = Q.get()

		pingtime = ping_addr(item[0])
		if pingtime < 0:
			pingtime = 0
		elif pingtime == 0:
			pingtime = 1

		if item[0] in IPs_bad.keys():
			if not pingtime:
				IPs_bad[item[0]][1] += 1
				IPs_bad[item[0]][2] = True
		elif not pingtime and mon_cli.get(item[1],False):
			IPs_bad[item[0]] = [mon_cli[item[1]],1,True]
		
		global Stats
		Stats.append([item[0],nowtime,pingtime])

		threads_processing_cnt[idx] += 1
		Q.task_done()

Q = Queue.Queue()
Stats = []
IPs_bad = {}
mon_cli = {}
threads_processing_cnt = []

def run_main():
	"""Главный поток. Получает блокировку, запускает дочерние потоки, генерирует
	для них очередь адресов, ждет выполнения, пишет в базу результат"""
	curtime = check_curtime()
	if not curtime:
		print "already processed"
		return 0

	cnt_ = 0

	if get_flock(curtime):
		if curtime % 288 == 240 and not __DEBUG_MODE: # ну бред же полный... не проще ли проверять по strftime("%H")==20? проще и нагляднее. но пока оставлю "шепикоходконем" в оригинале
			agg_1h(curtime)
		for i in range(__THREADS):
			t = threading.Thread(target=pinger_daemon,name="Thread #%i"%i,args=(i,curtime,time.strftime("%Y-%m-%d %H:%M:%S")))
			t.setDaemon(True)
			t.start()
			threads_processing_cnt.append(0)
		vals = get_source()
		for ipaddr in vals.keys():
			Q.put((ipaddr,vals[ipaddr]))
			cnt_+=1
		Q.join()

		res = set_result()
		cur = connection.cursor()
		cur.execute('start transaction')
		for i in res:
			if (__DEBUG_MODE and __DEBUG_SWAP_TABLES_MODE) or not __DEBUG_MODE:
				cur.execute(i)
			else:
				print i
		cur.execute('commit')
		cur.close()

		if (__DEBUG_MODE and __DEBUG_MAIL) or not __DEBUG_MODE:
			send_spam()

		release_flock(curtime)
	return cnt_

if __name__ == "__main__":
	print "BEGIN;#######################", time.strftime("%Y-%m-%d %H:%M:%S"), "#######################"
	cnt_ = run_main()
	for i in range(len(threads_processing_cnt)):
		print "Processed of Thread #%i - %i" % (i,threads_processing_cnt[i])
	print "Total Processed:", cnt_
	print "END;#########################", time.strftime("%Y-%m-%d %H:%M:%S"), "#######################\n\n"
